### PR TITLE
Extend foam model with solvent or water

### PR DESCRIPTION
**Pull request by totto82**
Depends on https://github.com/OPM/opm-common/pull/3523

### CODE REVIEW

File: opm/models/blackoil/blackoilfoammodules.hh [hunk 1]
```diff
@@ -475,8 +526,18 @@ public:
...
 
-        // adjust gas mobility
-        asImp_().mobility_[gasPhaseIdx] *= mobilityReductionFactor;
+        // adjust mobility
+        if (FoamModule::transportPhase() == Phase::WATER) {
+            asImp_().mobility_[waterPhaseIdx] *= mobilityReductionFactor;
+        } else if (FoamModule::transportPhase() == Phase::GAS) {
+            asImp_().mobility_[gasPhaseIdx] *= mobilityReductionFactor;
+        } else if (FoamModule::transportPhase() == Phase::SOLVENT) {
+            if constexpr (enableSolvent) {
```
Comment (GitPaean) [inline opm/models/blackoil/blackoilfoammodules.hh:535]: should we assert `enableSolvent == true` or throw if it is false. 

If we specify the transportPhase is SOLVENT, we should not let code continue running if `enableSolvent` is `false`, right?

File: opm/models/blackoil/blackoilfoammodules.hh [hunk 2]
```diff
@@ -475,8 +526,18 @@ public:
             mobilityReductionFactor = gasMobilityMultiplier.eval(foamConcentration_, /* extrapolate = */ true);
         }
 
-        // adjust gas mobility
-        asImp_().mobility_[gasPhaseIdx] *= mobilityReductionFactor;
+        // adjust mobility
+        if (FoamModule::transportPhase() == Phase::WATER) {
```
Comment (GitPaean) [inline opm/models/blackoil/blackoilfoammodules.hh:530]: better to use `switch` here.

File: opm/models/blackoil/blackoilfoammodules.hh [hunk 3]
```diff
@@ -280,23 +291,57 @@ public:
...
-            if (upIdx == inIdx) {
-                flux[contiFoamEqIdx] =
-                    extQuants.volumeFlux(gasPhaseIdx)
-                    *up.fluidState().invB(gasPhaseIdx)
-                    *up.foamConcentration();
+
+            // The effect of the mobility reduction factor is
+            // incorporated in the mobility for the relevant phase,
+            // so fluxes do not need modification here.
+            if (params_.transport_phase_ == Phase::WATER) {
```
Comment (GitPaean) [inline opm/models/blackoil/blackoilfoammodules.hh:299]: better to use `switch` here also, which is more readable.

**Comment by GitPaean** [comment]
The code looks fine to me. Do we have any example for water as the transport phase for foam or is it a future planned combination?

**Comment by totto82** [comment]
Thanks for the review. I have updated the code accordingly. 

> Do we have any example for water as the transport phase for foam or is it a future planned combination?

No, I am mostly interested in the SOLVENT part, but I though it made sense to extend the framework to also include water while working on it.

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
Thanks for the update. I am getting this PR in now.

**Pull request merged**